pkg/api/types.go

```diff
@@ -518,8 +518,18 @@ type GitRepoVolumeSource struct {
 // The contents of the target Secret's Data field will be presented in a volume
 // as files using the keys in the Data field as the file names.
 type SecretVolumeSource struct {
-	// Name of the secret in the pod's namespace to use
-	SecretName string `json:"secretName"`
+	// Name of the secret in the pod's namespace to use.  May not be specified if items is specified.
+	SecretName string `json:"secretName,omitempty"`
+	// A list of projections to make into the volume.  May not be specified if SecretName is specified.
+	Items []SecretVolumeFile `json:"items,omitempty"`
+}
+
+// SecretVolumeFile represents a single secret key projected into a file.
+type SecretVolumeFile struct {
+	// A reference to the secret key to project into the volume.
+	SecretKeySelector `json:",inline"`
```

Comment:
Alt model: a single "managed data volume source" that covers downward api, config map, and secret:

```
name: my-mega-dir
items:
  - configMapName: my-config-map
    files:
      - key: config
        file: myapp.conf
  - secretName: my-secret
    files:
      - key: password
        file: path/passwd.txt
  - dowardAPI: 
      apiVersion: v1
      fieldPath: metadata.name
    file: myapp.name.txt
```

I am not sure this is better, but it's an idea.